order initial implementation

Ansu-s opened the pull request.

rkpatra201 commented on src/main/resources/db/scripts/V7__ADDRESSTYPE_COLUMN.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE address
+    ADD COLUMN addressType VARCHAR(30) DEFAULT 'HOME';
```
This should bet INT type. See how we manage for BUY_NOW status.

rkpatra201 commented on src/main/resources/db/scripts/V8__CREATE_ORDER_TABLE.sql:
```diff
@@ -0,0 +1,8 @@
+    customer_id      BIGINT,
+    order_time       TIMESTAMP,
+    total_amount     DECIMAL(10, 2),
+    shipping_address VARCHAR(255)
```
use TEXT data type instead of VARCHAR (255) for address

rkpatra201 commented on src/main/java/com/example/bookstoreapp/filters/AuthenticationFilter.java:
```diff
@@ -19,7 +14,7 @@
 public class AuthenticationFilter implements Filter {
 
-  private static final String[] PUBLIC_URI_PATTERNS = {"/api/**", "/user/login", "/user/signup", "/health"};
+    private static final String[] PUBLIC_URI_PATTERNS = {"/**", "/api/**", "/user/login", "/user/signup", "/health"};
```
DONT apply /** .
Use login flow.

Ansu-s replied:
This is for development purpose I disabled the filter so I don't have to gone through auth process it will be removed after dev process ended this page was pushed unintendedly.

rkpatra201 replied:
disabling auth flow will break few things. such as we return userId from userContextService.
If you disable auth flow the userId will be returned null. which may cause null pointer exception.

Ansu-s replied:
Understood its value and will use the data from context during the order service implementation

rkpatra201 commented on src/main/resources/db/scripts/V7__ADDRESSTYPE_COLUMN.sql:
ADRRESSTYPE is not readable here. ADDRESS_TYPE is better.

rkpatra201 reviewed (changes requested):
please do changes as suggested

Ansu-s commented:
All changes will be added in the upcoming commit

rkpatra201 commented on src/main/java/com/example/bookstoreapp/entities/OrderEntity.java:
```diff
@@ -0,0 +1,22 @@
+public class OrderEntity {
+    @Id
+    private long order_id;
+    private long customer_id;
```
For java entity class here instead customer_id use customerId. In database it will be customer_id.
Same applies for other entity class fields.

rkpatra201 reviewed (changes requested):
requested changes

rkpatra201 commented on src/main/java/com/example/bookstoreapp/services/OrderService.java:
```diff
@@ -0,0 +1,13 @@
+public interface OrderService {
+     Order createOrder(Order order);
+
+     Order orderSummary(Long order_id);
```
orderId instead order_id

rkpatra201 reviewed (changes requested):
feedback given

Ansu-s commented:
- Added AddressType enum to int(address_type) conversation. 
- Fixed all given feedback

![Screenshot 2024-02-18 023146](https://github.com/learning-project-01/bookstore-app/assets/130679461/9c8abc93-7c1c-4b17-a99e-ea58cff13b35)  
![Screenshot 2024-02-18 023303](https://github.com/learning-project-01/bookstore-app/assets/130679461/8cf682b8-7228-4884-9d80-a1853625aad2)

rkpatra201 commented on src/main/java/com/example/bookstoreapp/filters/AuthenticationFilter.java:
```diff
@@ -19,62 +19,62 @@
 public class AuthenticationFilter implements Filter {
 
-  private static final String[] PUBLIC_URI_PATTERNS = {"/api/**", "/user/login", "/user/signup", "/health"};
+    private static final String[] PUBLIC_URI_PATTERNS = {"/api/**", "/user/login", "/user/signup", "/health"};
```
You should not commit filter changes here because your implementation does not need filter to be modified. Remove this change from your pull request l.

rkpatra201 commented on src/main/java/com/example/bookstoreapp/models/CartItemState.java:
```diff
@@ -4,29 +4,29 @@
-  IN_CART(0), // any SAVE_LATER must be IN_CART before BUY_NOW
-  SAVE_LATER(1), // it is in cart but in SAVE_LATER group
-  BUY_NOW(2); // inside cart select the cartItem to buy
+    IN_CART(0), // any SAVE_LATER must be IN_CART before BUY_NOW
```
This file doesn't need any modification. Don't put in your PR.

rkpatra201 commented on src/main/java/com/example/bookstoreapp/models/Order.java:
```diff
@@ -0,0 +1,36 @@
+
+@Data
+public class Order {
+    private long order_id;
```
Use orderId  instead order_id. This is java file and not a sql file.

rkpatra201 reviewed (changes requested):
Requested

rkpatra201 commented:
Test case is failing..Fix it.

Ansu-s commented:
## Order API
### createOrder   
Method : GET  
Endpoint : "/order"  
This will create a new order and set all the required fields for it and save into the database.
![image](https://github.com/learning-project-01/bookstore-app/assets/130679461/0f111f04-e472-4fb4-9b5c-7f9777d44671)
  
```
curl --location 'localhost:8080/order' \
--header 'X-AUTH-TOKEN: MDQ2ZWNkOTktNDBkOC00NTE0LWFkZmItNmNiOGVmMTdiZDE0' \
--data ''
```

### orderSummary
Method : GET  
Endpoint : "/order/{orderId}"  

This will show the specific order summary from the database by using the orderId.  
![image](https://github.com/learning-project-01/bookstore-app/assets/130679461/1dbb052e-c810-4419-9c6e-33bb609912a3)
```
curl --location 'localhost:8080/order/34646348435200' \
--header 'X-AUTH-TOKEN: MDQ2ZWNkOTktNDBkOC00NTE0LWFkZmItNmNiOGVmMTdiZDE0'
```

### allOrders  
Method : POST  
Endpoint : "/order/{customerId}"  
This will show all the list of orders a user has crated.  
![image](https://github.com/learning-project-01/bookstore-app/assets/130679461/2d33f4e2-a2df-46c2-9359-68d7ce6e6045)
```javascript
curl --location --request POST 'localhost:8080/order/273025944596600' \
--header 'X-AUTH-TOKEN: MDQ2ZWNkOTktNDBkOC00NTE0LWFkZmItNmNiOGVmMTdiZDE0'
```

rkpatra201 reviewed (changes requested):
Request

rkpatra201 commented on src/main/java/com/example/bookstoreapp/services/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+public class OrderServiceImpl implements OrderService {
+
+    @Autowired
+    UserContextService userContextService;
```
Use access modifier as private for the fields.

rkpatra201 commented on src/main/java/com/example/bookstoreapp/services/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+
+        if (addressList != null) {
+            try {
+                String jsonAddress = objectMapper.writeValueAsString(addressList.get(0));
```
What if I have 10 addresses, but here we are considering only the first address?

rkpatra201 commented on src/main/java/com/example/bookstoreapp/services/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+        } else throw new AppRuntimeException("No address exists for the user");
+
+        OrderEntity orderEntity = order.toEntity();
+        orderEntity = orderEntityRepository.save(orderEntity);
```
Order placed here but units not reduced in catalog.

rkpatra201 reviewed (changes requested):
Requested

rkpatra201 commented on src/main/java/com/example/bookstoreapp/models/AddressType.java:
```diff
@@ -0,0 +1,33 @@
+    HOME(0), // this is home address and it will be  default
+    WORK(1);  // this is for work address
+
+    private static Map<Integer, AddressType> weighVsType = new HashMap<>();
```
weight is correct spelling

rkpatra201 reviewed (changes requested):
Requested

Ansu-s commented:
Address Resolve issue fixed , now we can send a specific address id and get the shippingAddress changed to it
![Screenshot 2024-02-19 145056](https://github.com/learning-project-01/bookstore-app/assets/130679461/c08c6493-51d1-4aec-92ef-47c861b23e5c)
![Screenshot 2024-02-19 145158](https://github.com/learning-project-01/bookstore-app/assets/130679461/fea93364-0acd-4079-a402-f6e5a677aafc)

Ansu-s commented:
I have added the address checking for the order   
Created a new order :  

![image](https://github.com/learning-project-01/bookstore-app/assets/130679461/6446fc9a-ba38-4eb1-b297-fbe03857343e)  
orderSummary of a particular order :  

![image](https://github.com/learning-project-01/bookstore-app/assets/130679461/55fcadc5-b1ef-431b-b6e8-1c989150c542)  
All orders of a particular user  
![image](https://github.com/learning-project-01/bookstore-app/assets/130679461/4ee2f9be-31d2-436f-a0b9-1f6b3aaad8da)

rkpatra201 commented:
Accepted
